File: add_numbers.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

    def __repr__(self):
        out = []
        curr = self
        while curr:
            out.append(curr.val)
            curr = curr.next
        return str(out)


def list_to_linkedlist(li: list):
    start = ListNode(li[0])
    c = start
    for v in li[1:]:
        c.next = ListNode(v)
        c = c.next
    return start


def add_two_numbers(l1: ListNode, l2: ListNode):
    start = ListNode(0)
    p, q, curr = l1, l2, start
    carry = 0
    while p or q:
        x = p.val if p else 0
        y = q.val if q else 0
        addition = x + y + carry
        carry = addition // 10
        curr.next = ListNode(addition % 10)
        curr = curr.next
        if p:
            p = p.next
        if q:
            q = q.next

    if carry:
        curr.next = ListNode(carry)

    return start.next


def main():
    x = list_to_linkedlist([9, 9, 9, 9, 9, 9, 9])
    y = list_to_linkedlist([9, 9, 9, 9])

    print(f"{add_two_numbers(x, y) = }")


if __name__ == "__main__":
    main()

File: JavaScript/add_numbers.js
class ListNode {
    constructor(val, next) {
        this.val = (val === undefined ? 0 : val);
        this.next = (next === undefined ? null : next);
    }

    printList() {
        var out = [];
        var curr = this;
        while (curr != null) {
            out.push(curr.val);
            curr = curr.next;
        }
        return out.toString();
    }
}

var listToLinkedList = function (li) {
    var start = new ListNode(li[0]);
    c = start;
    li.slice(1).forEach(v => {
        c.next = new ListNode(v)
        c = c.next
    });
    return start
}

var addTwoNumbers = function (l1, l2) {
    var start = new ListNode(0);
    var [p, q, curr] = [l1, l2, start];
    var carry = 0;
    while (p !== null || q !== null) {
        var x = (p !== null ? p.val : 0);
        var y = (q !== null ? q.val : 0);
        var addition = x + y + carry;
        var carry = Math.floor(addition / 10);
        curr.next = new ListNode(addition % 10);
        curr = curr.next;
        if (p != null) {
            p = p.next;
        }
        if (q != null) {
            q = q.next;
        }
    }

    if (carry > 0) {
        curr.next = new ListNode(carry);
    }

    return start.next;
};

const a = listToLinkedList([9, 9, 9, 9, 9, 9, 9])
const b = listToLinkedList([9, 9, 9, 9])
console.log(addTwoNumbers(a, b).printList());